decision/blockstoremanager.go

```diff
@@ -0,0 +1,114 @@
+package decision
+
+import (
+	"context"
+	"sync"
+
+	blocks "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	bstore "github.com/ipfs/go-ipfs-blockstore"
+	process "github.com/jbenet/goprocess"
+)
+
+// blockstoreManager maintains a pool of workers that make requests to the blockstore.
+type blockstoreManager struct {
+	bs          bstore.Blockstore
+	workerCount int
+	jobs        chan func()
+}
+
+// newBlockstoreManager creates a new blockstoreManager with the given context
+// and number of workers
+func newBlockstoreManager(ctx context.Context, bs bstore.Blockstore, workerCount int) *blockstoreManager {
+	return &blockstoreManager{
+		bs:          bs,
+		workerCount: workerCount,
+		jobs:        make(chan func(), 1),
+	}
+}
+
+func (bsm *blockstoreManager) start(ctx context.Context, px process.Process) {
+	// Start up workers
+	for i := 0; i < bsm.workerCount; i++ {
+		i := i
+		px.Go(func(px process.Process) {
+			bsm.worker(ctx, i)
+		})
+	}
+}
+
+func (bsm *blockstoreManager) worker(ctx context.Context, id int) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case job := <-bsm.jobs:
+			job()
+		}
+	}
+}
+
+func (bsm *blockstoreManager) addJob(job func()) {
+	bsm.jobs <- job
+}
+
+func (bsm *blockstoreManager) getBlockSizes(ks []cid.Cid) map[cid.Cid]int {
+	res := make(map[cid.Cid]int)
+	if len(ks) == 0 {
+		return res
+	}
+
+	var lk sync.Mutex
+	bsm.jobPerKey(ks, func(c cid.Cid) {
+		size, err := bsm.bs.GetSize(c)
+		if err != nil {
+			if err != bstore.ErrNotFound {
+				log.Warningf("blockstore.GetSize(%s) error: %s", c, err)
```

Comment:
Is this really a warning?

Reply:
You mean it should be Error instead? Error seems right, I just want to make sure that's what you mean.

Reply:
I mean, not even a warning. Won't this happen every time we get a request for something we don't have?

Reply:
Oh, no this is when there's an error that's *not* `ErrNotFound` :)

Reply:
Oh! No, that's _totally_ an error and should be logged as such.